docs/datatypes/test/assertion.md

```diff
@@ -0,0 +1,497 @@
+
+Assume we have a function that concatenates two strings. One simple property of this function would be "the sum of the length of all inputs should be equal to the length of the output". Let's see an example of how we can make an assertion about this property:
+
+```scala mdoc:compile-only
+import zio.test._
+
+test("The sum of the lengths of both inputs must equal the length of the output") {
+  check(Gen.string, Gen.string) { (a, b) =>
+    assert((a + b).length)(Assertion.equalTo(a.length + b.length))
+  }
+}
+```
+
+The syntax of assertion in the above code, is `assert(expression)(assertion)`. The first section is an expression of type `A` which is _result_ of our computation and the second one is the expected assertion of type `Assertion[A]`.
+
+## Using Assertions with ZIO Tests
+
+We have two methods for writing test assertions:
+
+1. **`assert`** and **`assertM`**
+2. **`assertTrue`**
+
+The first one is the old way of asserting ordinary values and also ZIO effects. The second method, which is called _smart assertion_, has a unified logic for testing both ordinary values and ZIO effects. We encourage developers to use the smart assertion method, which is much simpler.
+
+### Classic Old-fashioned Assertions
+
+The `assert` and its effectful counterpart `assertM` are the old way of asserting ordinary values and ZIO effects.
+
+1. In order to test ordinary values, we should use `assert`, like the example below:
+
+```scala mdoc:compile-only
+import zio._
+import zio.test.{test, _}
+
+test("sum") {
+  assert(1 + 1)(Assertion.equalTo(2))
+}
+```
+
+2. If we are testing an effect, we should use the `assertM` function:
+
+```scala mdoc:compile-only
+import zio._
+import zio.test.{test, _}
+
+test("updating ref") {
+  val value = for {
+    r <- Ref.make(0)
+    _ <- r.update(_ + 1)
+    v <- r.get
+  } yield v
+  assertM(value)(Assertion.equalTo(1))
+}
+```
+
+3. Having this all in mind, probably the most common and also most readable way of structuring tests is to pass a for-comprehension to `test` function and yield a call to `assert` function.
+
+```scala mdoc:compile-only
+import zio._
+import zio.test.{test, _}
+
+test("updating ref") {
+  for {
+    r <- Ref.make(0)
+    _ <- r.update(_ + 1)
+    v <- r.get
+  } yield assert(v)(Assertion.equalTo(v))
+} 
+```
+
+### Smart Assertions
+
+The smart assertion is a simpler way to assert both ordinary values and effectful values. It uses the `assertTrue` function, which uses macro under the hood.
+
+1. Testing ordinary values:
+
+```scala mdoc:compile-only
+import zio._
+import zio.test.{test, _}
+
+test("sum"){
+  assertTrue(1 + 1 == 2)
+}
+```
+
+2. Testing effectful values:
+
+```scala mdoc:compile-only
+import zio._
+import zio.test.{test, _}
+
+test("updating ref") {
+  for {
+    r <- Ref.make(0)
+    _ <- r.update(_ + 1)
+    v <- r.get
+  } yield assertTrue(v == 1)
+}
+```
+
+## Logical Operations
```

Comment:
There is a lot of great content here and below that would probably have helped people a lot with the old assertions. Maybe this can live in the 1.0 branch but would focus this more on the smart assertions and how you do the same thing with them.

Reply:
The `zio.mock.Expectation` relies on the old assertions. 
After the migration to the new assertion style, I will revise this page.